### PR TITLE
Implement adde-modal using ember-modal-dialog

### PR DESCRIPTION
**Pull request by billylittlefield**
# adde-modal

This change-set implements the `{{adde-modal}}` block component, which is essentially a skinned wrapper around [ember-dialog-modal@1.0.0](https://github.com/yapplabs/ember-modal-dialog/tree/3b07c8db7ab9972796507db076c337c2a5c052fe), which supports from Ember 1.10 <=> Ember 2.4. Under the hood, `ember-dialog-modal` uses `ember-wormhole` to bring the rendered modal higher in the DOM, just underneath `.ember-application`. Using this add-on prevents us from reinventing the wheel, and also gives us an easy upgrade path as consumers of this wrapper (ie, Iverson) are updated to Ember 2.0 and beyond.

## Implementation

The goal of adde-modal was to harness the power of an existing solution for modals that is "future-proofed", then apply our own styling / functionality as needed within Addepar's application. This should hopefully make replacing existing ember-widgets modals with adde-modals a pain-free experience, because we have feature parity. Namely, this meant incorporating the following functionality:

- styling (handled in a [separate PR](https://github.com/Addepar/addepar-style-toolbox/pull/10) on `@addepar/style-toolbox`)
- modal header with customizable title and (optional) X-button to close
- (optional) modal footer with (optional) Confirm and Cancel buttons
  - customizable button text for these buttons
- options to close modal by clicking outside the modal and/or pressing 'Escape'
- three built-in size options (small, normal, large)

The major difference between the ember-widgets modal and this new `adde-modal` is how they are popped up. The old ember-widgets modal is shown by calling `.popup(opts)` on the component class in a JavaScript file. The `adde-modal` behaves the same way as `ember-dialog-modal`, and is actually put into a template directly. It should be wrapped in a conditional `{{#if}}` block using a variable that controls when the modal is shown.

**Example:**

```hbs
{{!-- component template --}}
{{#if showModal}}
  {{#adde-modal
    onClose='closeModal'
  }}
    Content inside the modal
  {{/adde-modal}}
{{/if}}
```
```js
// component javascript
export default Component.extend({
  showModal: false,
  actions: {
    openModal() {
      this.set('showModal', true);
    },
    closeModal() {
      this.set('showModal', false);
    }
  }
})
```

There are certain existing modal implementations in Iverson that deviate from the standard footer, and customizing just the button texts / existence isn't enough. When full customization of the modal is needed, you can use `{{adde-modal-container}}` to customize everything. Example:

```hbs
{{#if isShowingCustomModal}}
  {{#adde-modal-container}}

    {{adde-modal-header headerText="My Custom Modal"}}

    {{#adde-modal-body}}
      Custom body content
    {{/adde-modal-body}}

    {{#adde-modal-footer}}
      <div {{action "customAction"}}>Custom footer content with custom actions</div>
    {{/adde-modal-footer}}

  {{/adde-modal-container}}
{{/if}}
```

The `adde-modal` directly extends the `ModalDialog` component from `ember-modal-dialog`, overriding certain default arguments and providing new action handlers for confirm, cancel, close, and clicking the overlay. In order to add the modal header / footer / buttons without nesting the modal , it uses its own layout modified directly from the *.hbs for the `ModalDialog` component.

I opted to use ember-decorators and ES6 class syntax, as well as the formatting / linting rules from `@addepar/ember-toolbox`, which then necessitated the inclusion of `eslint-babel` to green-light the experimental ES6 feature use.

## Functionality
Generally speaking, all the same options in `ember-modal-dialog` can be passed into the `adde-modal` component, with some new options. The full list of options is:

| Property | Purpose |
| --- | --- |
`headerText`          | Text to be shown in the header of the modal.
`translucentOverlay`  | Indicates translucence of overlay, toggles presence of `translucent` CSS selector. Defaults to `true`.
`confirmText`         | Button text for the 'Confirm' button. Defaults to 'Confirm'. If set to null, the 'Confirm' button will not be shown.
`onConfirm`           | Action sent to parent when the 'Confirm' button is clicked.
`cancelText`          | Button text for the 'Cancel' button. Defaults to 'Cancel'. If set to null, the 'Cancel' button will not be shown.
`onCancel`            | Action sent to parent when the 'Cancel' button is clicked.
`onClickOverlay`      | Action sent to parent when the translucent overlay behind the modal is clicked.
`onClose`             | Action sent to parent when the 'Close' button is clicked. This is also the fallback action if action names are not defined for `onConfirm`, `onCancel`, or `onClickOverlay`
`showFooter`          | Whether to show the modal footer, which contains the 'Confirm' and 'Cancel' buttons (if used).
`showCloseButton`     | Whether to show the close button in the modal header. Defaults to true.
`escapeToClose`       | Whether pressing the escape key should close the modal. Defaults to true.
`clickOverlayToClose` | Indicates whether clicking outside a modal should close the modal. Defaults to true.
`containerClassNames` | CSS class names to append to container divs. This is a concatenated property, so it does **not** replace the default container class (default: `'ember-modal-dialog'`. If you subclass this component, you may define this in your subclass.)
`overlayClassNames`   | CSS class names to append to overlay divs. This is a concatenated property, so it does **not** replace the default overlay class (default: `'ember-modal-overlay'`. If you subclass this component, you may define this in your subclass.)
`wrapperClassNames`   | CSS class names to append to wrapper divs. This is a concatenated property, so it does **not** replace the default container class (default: `'ember-modal-wrapper'`. If you subclass this component, you may define this in your subclass.)
`target`              | Element selector, element, or Ember View reference that serves as the reference for modal position (default: `'body'`)
`renderInPlace`       | When true, renders the modal without wormholing, useful for including a modal in a style guide
`targetAttachment`    | A string of the form 'vert-attachment horiz-attachment', e.g. `'middle left'` (see "Positioning" section below)

## Accessibility

This base PR includes accessibility support for the modal, including:

- `role='dialog'` on modal element
- `aria-modal='true'` on modal element
- `aria-labelledBy='[MODAL_HEADER_ID]'` on modal element
- initial focus set to: Confirm button, Cancel button, Close button, first focusable element, none (listed in order of priority)
- close modal on 'Escape' enabled by default

## Tests

Tests are still needed. I'm happy to either add them to this PR, or open a different PR once they're finished.

## Screenshot

image

If you want to pull this branch to play around locally, you'll need to either `yarn link` the `@addepar/style-toolbox` to [this PR](https://github.com/Addepar/addepar-style-toolbox/pull/10), or update the commit hash in the `package.json`

### CODE REVIEW

File: addon/components/adde-modal.js [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    this._rootElement.removeEventListener('keydown', this._closeModalHandler);
+  }
+
+  // ----- Private Functions -----
+
+  /**
+   * Returns a node list of all focusable elements in the modal
+   * @returns {NodeList}
+   */
+  _getFocusableElementsInPopper() {
```
Comment (LydiaBarnes) [inline addon/components/adde-modal.js]: @pzuraq now that we need this in more than one place... 🤔

File: addon/templates/components/adde-modal.hbs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }}
+      {{#ember-modal-dialog-positioned-container 
+        classNameBindings='containerClassNamesString targetAttachmentClass container-class sizeClass'
+        classNames='adde-modal'
+        targetAttachment=targetAttachment
+        target=target
+      }}
+        <div class='adde-modal-header'>
+          {{#if showCloseButton}}
+            <button tabIndex=0 class='button-icon adde-modal-close' {{action 'onClose'}}>
```
Comment (LydiaBarnes) [inline addon/templates/components/adde-modal.hbs]: dont think this really matters, but its usually `tabindex`

**Review by LydiaBarnes — COMMENTED** [review]
Looks like a good start!
Wondering about the footer, and how we'd modify the buttons. We'll prob have a few modal types with predetermined buttons (like dialog with one Ok button for example), but there is also a need for a custom footer.

**Comment by billylittlefield** [comment]
@LydiaBarnes 
In the footer, there can be 2 buttons (one primary, one default) that can both be conditionally rendered and text adjusted. So, for instance, if we want a dialog modal with just a 'OK' in default button style, we could not show the 'Confirm' button, show the 'Cancel' button, and change the 'Cancel' button text to 'OK'.

The remainder of the footer content is not customizable as you said, but I'm wondering what the use case for customizing that is? The current footer in the ember-widgets modal only allows button customization as well IIRC

**Comment by LydiaBarnes** [comment]
In terms of the base modal components that should be fine. I'm thinking ahead about all the modals in Iverson where we used a custom footer bc the base buttons weren't quite what we wanted.
I don't want to over complicated it now. Suppose we can get to that when we start replacing modals in Iverson, come across any issues and see what we think.

**Comment by billylittlefield** [comment]
Ah you're right, I forgot we supported customizing the footer view with `footerViewClass`. Looks like we have 7 modals doing this now, most of which have relatively minor changes to the footer (adding loading spinner, disabling buttons, adding a checkbox, page numbers, etc). I'll think about this now.

**Pull request merged**